### PR TITLE
implement status condition transitions for vertica restore points query cr

### PR DESCRIPTION
**Pull request by chinhtranvan**
In the VerticaRestorePointsQuery CRD we have status conditions to know when the query has started and when it has finished.

  conditions:
  - lastTransitionTime: <ts>
    status: "True"
    type: Querying
  - lastTransitionTime: <ts>
    status: "True"
    type: QueryComplete

This PR can add the status condition logic as follows:
- skip reconcile if QueryComplete is true
- set Querying status condition prior to calling vclusterops API
- clear Querying status condition after calling API. Include an error message in the status condition if the API fails.
- set the QueryComplete if the vclusterops API suceeded

### CODE REVIEW

File: api/v1beta1/verticarestorepointsquery_types.go [hunk 1]
```diff
@@ -47,8 +48,54 @@ const (
 // VerticaRestorePointsQueryStatus defines the observed state of VerticaRestorePointsQuery
 type VerticaRestorePointsQueryStatus struct {
 	// +operator-sdk:csv:customresourcedefinitions:type=status
-	// RestorePoints used to list out the available restore points.
-	RestorePoints string `json:"restorePoints"`
+	// Conditions for VerticaRestorePointsQuery
+	Conditions []VerticaRestorePointsQueryCondition `json:"conditions,omitempty"`
```
Comment (spilchen) [inline api/v1beta1/verticarestorepointsquery_types.go:52]: We should use the `metav1.Condition` type rather than creating our own. This allows us to use the condition functions in the operator-sdk package. We use this in the v1 VerticaDB API (search for metav1.Condition).

File: pkg/controllers/vrpq/verticarestorepointsquery_controller.go [hunk 2]
```diff
@@ -102,9 +102,11 @@ func (r *VerticaRestorePointsQueryReconciler) SetupWithManager(mgr ctrl.Manager)
 // constructActors will a list of actors that should be run for the reconcile.
 // Order matters in that some actors depend on the successeful execution of
 // earlier ones.
-func (r *VerticaRestorePointsQueryReconciler) constructActors(_ *vapi.VerticaRestorePointsQuery,
-	_ logr.Logger) []controllers.ReconcileActor {
+func constructActors(r *VerticaRestorePointsQueryReconciler, vrpq *vapi.VerticaRestorePointsQuery,
```
Comment (spilchen) [inline pkg/controllers/vrpq/verticarestorepointsquery_controller.go:105]: Why did you move this function out of the `VerticaRestorePointsQueryReconciler` struct?

File: pkg/vadmin/restore_points_query_at.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+*/
+
+package vadmin
+
+import (
+	"context"
+)
+
+func (a *Admintools) ListRestorePoints(_ context.Context) error {
+	a.Log.Info("start restore points query")
```
Comment (spilchen) [inline pkg/vadmin/restore_points_query_at.go:23]: I don't think we will ever implement this in the admintools side. Lets return an error if this is ever called.

File: pkg/controllers/vrpq/listrestorepoints_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/go-logr/logr"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1beta1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	vrpqstatus "github.com/vertica/vertica-kubernetes/pkg/vrpqstatus"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type ListRestorePointsQueryReconciler struct {
```
Comment (spilchen) [inline pkg/controllers/vrpq/listrestorepoints_reconciler.go:30]: We kind of repeat the name in the package. Maybe just call it the `QueryReconciler`?

File: pkg/controllers/vrpq/listrestorepoints_reconciler.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		v.Vrpq.Status.Conditions[vapi.QueryCompleteIndex].Status == corev1.ConditionTrue {
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, v.setListRestorePointsQueryConditions(ctx, req)
+}
+
+// setListRestorePointsQueryConditions will update the status condition before and after calling
+// list restore points api
+func (v *ListRestorePointsQueryReconciler) setListRestorePointsQueryConditions(ctx context.Context, _ *ctrl.Request) error {
```
Comment (spilchen) [inline pkg/controllers/vrpq/listrestorepoints_reconciler.go:58]: We should have a unit test for this reconciler.

**Review by spilchen — COMMENTED** [review]
We should include a simple e2e test. Even just to wait on the new status condition after creating the CR.

File: pkg/controllers/vrpq/verticarestorepointsquery_controller.go (hunk 2 above)
Comment (chinhtranvan) [reply pkg/controllers/vrpq/verticarestorepointsquery_controller.go]: oops, Will move it back

File: pkg/controllers/vrpq/listrestorepoints_reconciler_test.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	vapi "github.com/vertica/vertica-kubernetes/api/v1beta1"
+	"github.com/vertica/vertica-kubernetes/pkg/cmds"
+	vmeta "github.com/vertica/vertica-kubernetes/pkg/meta"
+	"github.com/vertica/vertica-kubernetes/pkg/test"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const TestPassword = "test-pw"
+
+var _ = Describe("stopdb_reconcile", func() {
```
Comment (spilchen) [inline pkg/controllers/vrpq/listrestorepoints_reconciler_test.go:33]: copy/paste error?
```suggestion
var _ = Describe("query_reconciler", func() {
```

File: pkg/vadmin/restore_points_query_at.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+package vadmin
+
+import (
+	"context"
+	"errors"
+)
+
+func (a *Admintools) ListRestorePoints(_ context.Context) error {
+	return errors.New("ListRestorePoints is not supported in the admintools side")
```
Comment (spilchen) [inline pkg/vadmin/restore_points_query_at.go:24]:
```suggestion
	return errors.New("ListRestorePoints is not supported for admintools deployments")
```

File: pkg/vrpqstatus/status.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func Update(ctx context.Context, clnt client.Client, log logr.Logger, vrpq *vapi.VerticaRestorePointsQuery,
+	updateFunc func(*vapi.VerticaRestorePointsQuery) error) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		// Always fetch the latest to minimize the chance of getting a conflict error.
+		nm := types.NamespacedName{Namespace: vrpq.Namespace, Name: vrpq.Name}
+		if err := clnt.Get(ctx, nm, vrpq); err != nil {
+			log.Info("VerticaRestorePointsQuery resource not found.  Ignoring since object must be deleted")
```
Comment (spilchen) [inline pkg/vrpqstatus/status.go:37]: This should be done if the IsNotFound() function returned true. We can't assume all errors are not found errors.

File: pkg/controllers/vrpq/listrestorepoints_reconciler.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+		v1.MakeCondition(vapi.Querying, metav1.ConditionTrue, ""))
+	if err != nil {
+		return err
+	}
+
+	errAPI := q.Dispatcher.ListRestorePoints(ctx)
+	// Include an error message in the status condition if the API fails.
+	if errAPI != nil {
+		q.VRec.Log.Info("Fail to call Vcluster list restore points API")
+		return errAPI
```
[the remark this answers is not in the document]
Comment (chinhtranvan) [reply pkg/controllers/vrpq/listrestorepoints_reconciler.go:71]: If we want to clear the Querying status condition before we exit, I believe we should move lines 74-79 before the API call fails, as we want to clear it regardless of the API result.

**Comment by chinhtranvan** [comment]
> We should include a simple e2e test. Even just to wait on the new status condition after creating the CR.

@spilchen I apply a manifest yaml file 
```
apiVersion: vertica.com/v1beta1
kind: VerticaRestorePointsQuery
metadata:
  name: archive-query
spec:
  verticaDBName: vert
  archiveName: backup 

```
Then I use ` k describe VerticaRestorePointsQuery archive-query` and get this info 
```
Name:         archive-query
Namespace:    test
Labels:       <none>
Annotations:  <none>
API Version:  vertica.com/v1beta1
Kind:         VerticaRestorePointsQuery
Metadata:
  Creation Timestamp:  2024-01-10T04:34:57Z
  Generation:          1
  Resource Version:    14783
  UID:                 77670a09-d6e6-4fcd-b298-00d4ba0a6803
Spec:
  Archive Name:     backup
  Vertica DB Name:  vert
Status:
  Conditions:
    Last Transition Time:  2024-01-10T04:35:02Z
    Message:               
    Reason:                UnKnown
    Status:                True
    Type:                  Querying
    Last Transition Time:  2024-01-10T04:35:02Z
    Message:               
    Reason:                UnKnown
    Status:                True
    Type:                  QueryComplete
Events:                    <none>
```
I will open another Jira to add this simple e2e tests, since my brain doesn't work well to figure out the way to add this e2e test.

File: pkg/controllers/vrpq/query_reconciler.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	// set Querying status condition prior to calling vclusterops API
+	err := vrpqstatus.UpdateCondition(ctx, q.VRec.Client, q.VRec.Log, q.Vrpq,
+		v1.MakeCondition(vapi.Querying, metav1.ConditionTrue, ""))
+	if err != nil {
+		return err
+	}
+
+	// clear Querying status condition
+	err = vrpqstatus.UpdateCondition(ctx, q.VRec.Client, q.VRec.Log, q.Vrpq,
+		v1.MakeCondition(vapi.Querying, metav1.ConditionUnknown, ""))
```
Comment (spilchen) [inline pkg/controllers/vrpq/query_reconciler.go:68]: Any reason why we didn't pick `metav1.ConditionalFalse` to turn off the status condition.  Unknown is reserved for "k8s can't decide if a resource is in the condition or not"

File: pkg/controllers/vrpq/query_reconciler.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+// Temporarily, runListRestorePoints will not call the ListRestorePoints API
+// since the dispatcher is not set up yet
+func (q *QueryReconciler) runListRestorePoints(ctx context.Context, _ *ctrl.Request) error {
+	// set Querying status condition prior to calling vclusterops API
+	err := vrpqstatus.UpdateCondition(ctx, q.VRec.Client, q.VRec.Log, q.Vrpq,
+		v1.MakeCondition(vapi.Querying, metav1.ConditionTrue, ""))
+	if err != nil {
+		return err
+	}
+
```
Comment (spilchen) [inline pkg/controllers/vrpq/query_reconciler.go:65]: Maybe add a comment to indicate the API call should be done here

File: pkg/controllers/vrpq/query_reconciler.go (hunk 10 above)
Comment (chinhtranvan) [reply pkg/controllers/vrpq/query_reconciler.go:68]: yes, will update to `metav1.ConditionalFalse`

**Comment by spilchen** [comment]
Latest changes look good. As discussed in scrum, lets add the e2e test to this PR.

File: tests/e2e-leg-6/verify-vrpq-conditions/setup-vrpq/base/setup-vrpq.yaml
Comment (spilchen) [inline tests/e2e-leg-6/verify-vrpq-conditions/setup-vrpq/base/setup-vrpq.yaml:1]: I don't think we need kustomize for this. We don't do any kind of substitution. Lets put this CR right in step 3 and avoid making the kustomize call.

File: tests/e2e-leg-6/verify-vrpq-conditions/03-setup.yaml [hunk 12]
```diff
@@ -11,7 +11,15 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-apiVersion: kuttl.dev/v1betat1
+apiVersion: kuttl.dev/v1beta1
 kind: TestStep
 commands:
-  - command: bash -c "kustomize build setup-vrpq/overlay | kubectl -n $NAMESPACE apply -f - "
+  - script: |
```
Comment (spilchen) [inline tests/e2e-leg-6/verify-vrpq-conditions/03-setup.yaml:17]: You don't need to use kubectl here. Just replace the kuttl.dev/v1beta1 YAML manifest with your new CR.

**Review by spilchen — APPROVED** [review]
